Описание
В данном блокноте исследуется распределение уникальных идентификаторов клиентов (cn) и пользователей (user_id) по странам, основываясь на данных, полученных из client_ip. Также проводится анализ доли клиентов и пользователей, взаимодействующих с сервисом, основанным на наличии домена bablosoft в запросах (req_host).

Этапы
Подключение к ClickHouse: Настраивается безопасное соединение с базой данных zprx с использованием учетных данных из переменных окружения.

Формирование первого запроса:

Получение данных о cn и соответствующих странах с фильтрацией:
Период: август 2024 года.
Исключение запросов из России.
Исключение ошибочных запросов.
Выделение cn, связанного с сервисом bablosoft.
Формирование второго запроса:

Извлечение всех уникальных связей между user_id и cn из таблицы zprx_ds.device.
Объединение данных:

Объединение двух выборок по cn для анализа связей с учетом стран и пользователей.
Анализ данных:

Подсчет уникальных значений cn и user_id для каждой страны.
Вычисление доли идентификаторов (cn и user_id), связанных с сервисом bablosoft.
Формирование итоговой таблицы с результатами для топ-40 стран.

In [1]:
#Скрипт используется для поиска сразну нескольких сервисов: bablosoft, ipinfo, zennolab
#Для того, чтобы найти процент cn с сервисом по странам, достаточно подствить эти значеня в метод LIKE в запросе SQL

In [2]:
import os
from clickhouse_driver import Client
import pandas as pd

# Получаем пароль из переменной окружения
password = os.getenv("CLICKHOUSE_PASSWORD")

# Проверяем, что пароль не является None
if password is None:
    raise ValueError("Переменная окружения CLICKHOUSE_PASSWORD не установлена")

client = Client(
    host='ch.infra.zprx.net',
    port=2044,
    user='zprx_analytics',
    password=password,
    secure=True,
    verify=False,
    database='zprx'
)

try:
    query = """
SELECT cn,
       CASE 
           WHEN req_host LIKE '%bablosoft%' THEN cn 
           ELSE NULL 
       END AS cn_with_service,
       geoip2_country(client_ip) country
FROM 
     zprx.zproxy_reqs_v1__merge
WHERE toYYYYMM(ts_millis) = toYYYYMM(toDate('2024-08-01'))
  AND geoip2_country(client_ip) != 'Russia'
  AND error = '0' 
GROUP BY cn, 
         cn_with_service,
         country
    """

    # Execute the query and store the result in DataFrame
    result = client.execute(query)
    country_per_cn = pd.DataFrame(result, columns=["cn", "cn_with_service", "country"])

    # Print the results in table format
    print(country_per_cn)

finally:
    client.disconnect()

               cn cn_with_service         country
0      eyoIMBbFZD            None          Poland
1      nNKe5pN3AO            None  United Kingdom
2      JUCo9VlFzy            None         Finland
3      bkYvoHgxmr            None       Hong Kong
4      9uTHcpBvez            None           Chile
...           ...             ...             ...
30207  CoyDahffZl            None     Netherlands
30208  GgFeVgu47I            None                
30209  yZZ7Wx7ACX            None         Finland
30210  PBM0bMK61O            None         Germany
30211  kyRqTvda5C            None       Singapore

[30212 rows x 3 columns]


In [3]:
import os
from clickhouse_driver import Client
import pandas as pd

# Получаем пароль из переменной окружения
password = os.getenv("CLICKHOUSE_PASSWORD")

# Проверяем, что пароль не является None
if password is None:
    raise ValueError("Переменная окружения CLICKHOUSE_PASSWORD не установлена")

client = Client(
    host='ch.infra.zprx.net',
    port=2044,
    user='zprx_analytics',
    password=password,
    secure=True,
    verify=False,
    database='zprx'
)

try:
    query = """
SELECT user_id,
       cn
FROM zprx_ds.device
WHERE user_id is not NUll       
GROUP BY user_id,
         cn
    """

    # Execute the query and store the result in DataFrame
    result = client.execute(query)
    country_per_user_id = pd.DataFrame(result, columns=["user_id", "cn"])

    # Print the results in table format
    print(country_per_user_id)

finally:
    client.disconnect()

          user_id          cn
0      rbwnJagmNN  Cb9hjRB849
1      EI38Acy1Fu  mW3r5vCPDq
2      SDRGfJMG0v  MbYBbZMMZ4
3      pOu6lPJB7O  JyW7Ax7yo2
4      kEVUaSXFTd  qKI4DRKVch
...           ...         ...
70290  NOP9RAwaNo  vlJbrXnnce
70291  MjOvWPJctn  VFEm6CHvos
70292  WckroSFUsK  IAGNs9LuqA
70293  Yzz1odAD28  p8trySDwOz
70294  XoJdMcvjbY  jEjgMeckvY

[70295 rows x 2 columns]


In [4]:
user_id_cn_countries = pd.merge(country_per_cn, country_per_user_id, on='cn', how='left')


In [5]:
# Подсчет общего числа уникальных значений 'cn' для каждой страны
unique_cn_per_country = user_id_cn_countries.groupby('country')['cn'].nunique().reset_index(name='total_unique_cn')  # Группировка по стране и подсчет уникальных 'cn'

# Подсчет уникальных значений 'cn', где 'cn_with_service' не None, для каждой страны
unique_cn_with_service_per_country = user_id_cn_countries[user_id_cn_countries['cn_with_service'].notna()].groupby('country')['cn'].nunique().reset_index(name='unique_cn_with_service')  # Фильтрация по 'cn_with_service' и подсчет уникальных 'cn'

# Объединение данных и замена NaN на 0
country_cn_service_share = pd.merge(unique_cn_per_country, unique_cn_with_service_per_country, on='country', how='left').fillna(0)  # Объединение по стране и заполнение пропусков нулями

# Расчет доли значений с учетом сервиса
country_cn_service_share['share_cn_with_service (%)'] = (country_cn_service_share['unique_cn_with_service'] / country_cn_service_share['total_unique_cn']) * 100  # Доля в процентах

# Приведение к целочисленному типу для удобства
country_cn_service_share['unique_cn_with_service'] = country_cn_service_share['unique_cn_with_service'].astype(int)  # Приведение к целым числам
country_cn_service_share['share_cn_with_service (%)'] = country_cn_service_share['share_cn_with_service (%)'].astype(int)  # Приведение к целым числам

# Сортировка по доле в убывающем порядке и выбор топ-40
country_cn_service_share.sort_values(by='share_cn_with_service (%)', ascending=False).head(40)  # Сортировка и вывод топ-40


country  total_unique_cn  unique_cn_with_service  \
131                Uruguay               81                      72   
25                Colombia              294                     187   
53               Indonesia              114                      32   
22                  Canada              316                      83   
90             Netherlands             1107                     258   
89                   Nepal               22                       4   
126                 Turkey              380                      65   
62                  Jordan                6                       1   
67              Kyrgyzstan                7                       1   
1              Afghanistan               19                       2   
127                Ukraine             1065                     110   
11              Bangladesh              164                      15   
106                Romania              290                      27   
42                 Germany             2119                     151   
112               Slovenia               26                       2   
129         United Kingdom              706                      46   
86                 Morocco               49                       3   
31      Dominican Republic               16                       1   
134                Vietnam              285                      18   
130          United States             2007                     119   
18                  Brazil              998                      56   
37                 Finland             2021                     112   
114            South Korea               55                       3   
52                   India              190                       8   
117                 Sweden              107                       5   
63              Kazakhstan              404                      19   
58                  Israel               93                       4   
35                 Estonia               57                       2   
6                  Armenia               27                       1   
97                Pakistan               72                       1   
38                  France              406                       7   
110              Singapore              577                      10   
115                  Spain              260                       5   
29                 Czechia              157                       3   
118            Switzerland              132                       2   
119                  Syria               64                       1   
14                 Belgium               67                       1   
81                 Mayotte                5                       0   
98   Palestinian Territory                1                       0   
83                 Moldova              151                       1   

     share_cn_with_service (%)  
131                         88  
25                          63  
53                          28  
22                          26  
90                          23  
89                          18  
126                         17  
62                          16  
67                          14  
1                           10  
127                         10  
11                           9  
106                          9  
42                           7  
112                          7  
129                          6  
86                           6  
31                           6  
134                          6  
130                          5  
18                           5  
37                           5  
114                          5  
52                           4  
117                          4  
63                           4  
58                           4  
35                           3  
6                            3  
97                           1  
38                           1  
110                          1  
115               

In [6]:
# Подсчет общего числа уникальных значений 'user_id' для каждой страны
unique_user_id_per_country = user_id_cn_countries.groupby('country')['user_id'].nunique().reset_index(name='total_unique_user_id')  # Группировка по стране и подсчет уникальных user_id

# Подсчет уникальных значений 'user_id', где 'cn_with_service' не None, для каждой страны
unique_user_id_with_service_per_country = user_id_cn_countries[user_id_cn_countries['cn_with_service'].notna()].groupby('country')['user_id'].nunique().reset_index(name='unique_user_id_with_service')  # Фильтрация по cn_with_service и подсчет уникальных user_id

# Объединение данных в одну таблицу
country_user_service_share = pd.merge(unique_user_id_per_country, unique_user_id_with_service_per_country, on='country', how='left').fillna(0)  # Объединение по стране и заполнение пропусков нулями

# Расчет доли user_id с учетом сервиса в процентах
country_user_service_share['share_user_id_with_service (%)'] = (country_user_service_share['unique_user_id_with_service'] / country_user_service_share['total_unique_user_id']) * 100  # Доля в процентах

# Приведение типов данных для удобного отображения
country_user_service_share['unique_user_id_with_service'] = country_user_service_share['unique_user_id_with_service'].astype(int)  # Приведение к целым числам
country_user_service_share['share_user_id_with_service (%)'] = country_user_service_share['share_user_id_with_service (%)'].astype(int)  # Приведение к целым числам

# Сортировка данных по убыванию доли и вывод топ-40 записей
country_user_service_share.sort_values(by='share_user_id_with_service (%)', ascending=False).head(40)  # Сортировка и выбор топ-40



country  total_unique_user_id  unique_user_id_with_service  \
112            Slovenia                     5                            2   
131             Uruguay                     6                            2   
1           Afghanistan                     4                            1   
62               Jordan                     4                            1   
126              Turkey                    83                           20   
90          Netherlands                   147                           27   
89                Nepal                     6                            1   
11           Bangladesh                    19                            3   
42              Germany                   242                           36   
67           Kyrgyzstan                     7                            1   
114         South Korea                    20                            2   
37              Finland                   233                           23   
6               Armenia                    11                            1   
63           Kazakhstan                    37                            3   
31   Dominican Republic                    12                            1   
127             Ukraine                   302                           23   
130       United States                   299                           22   
134             Vietnam                    54                            4   
53            Indonesia                    39                            3   
25             Colombia                    39                            3   
119               Syria                    13                            1   
97             Pakistan                    16                            1   
22               Canada                    59                            4   
52                India                    46                            3   
29              Czechia                    40                            2   
38               France                    96                            5   
35              Estonia                    23                            1   
115               Spain                    81                            4   
69               Latvia                    46                            2   
117              Sweden                    42                            2   
118         Switzerland                    47                            2   
14              Belgium                    25                            1   
129      United Kingdom                   142                            7   
86              Morocco                    21                            1   
58               Israel                    26                            1   
83              Moldova                    27                            1   
49            Hong Kong                    78                            2   
106             Romania                    67                            2   
59                Italy                    35                            1   
110           Singapore                    95                            2   

     share_user_id_with_service (%)  
112                              40  
131                              33  
1                                25  
62                               25  
126                              24  
90                               18  
89                               16  
11                               15  
42                               14  
67                               14  
114                              10  
37                                9  
6                                 9  
63                                8  
31                                8  
127                               7  
130                               7  
134                               7  
53                                7  
25                                7  
119                

In [7]:
country_cn_service_share.to_csv('доля_cn_с_bablosoft_по_странам(client_ip).csv', index=False)

In [8]:
country_user_service_share.to_csv('доля_user_id_с_bablosoft_по_странам(client_ip).csv', index=False)

Результаты
Для cn:

Вычислено общее количество уникальных cn и доля cn, связанных с сервисом.
Пример: Уругвай имеет 81 уникальный cn, из которых 72 связаны с bablosoft (88%).
Для user_id:

Вычислено общее количество уникальных user_id и доля тех, кто взаимодействовал с сервисом.
Пример: Словения имеет 5 уникальных user_id, из которых 2 связаны с bablosoft (40%).